plutus-core/plutus-core/src/PlutusCore/Parser/Builtin.hs

```diff
@@ -6,80 +6,34 @@ module PlutusCore.Parser.Builtin where
 import PlutusCore.Parser.Type (defaultUniType)
+import PlutusCore.Pretty (display)
+import Prelude hiding (lookup)
 
+mkCachedBuiltin :: [DefaultFun] -> Map T.Text DefaultFun -> Map T.Text DefaultFun
+mkCachedBuiltin (hdFns:tlFns) builtinMap =
+    mkCachedBuiltin tlFns (insert (display hdFns) hdFns builtinMap)
+mkCachedBuiltin [] builtinMap = builtinMap
 
--- | The list of parsable default functions and their pretty print correspondence.
-builtinFnList :: [(DefaultFun, T.Text)]
-builtinFnList =
-    [ (AddInteger,"addInteger")
-    , (SubtractInteger,"subtractInteger")
-    , (MultiplyInteger,"multiplyInteger")
-    , (DivideInteger,"divideInteger")
-    , (QuotientInteger,"quotientInteger")
-    , (RemainderInteger,"remainderInteger")
-    , (ModInteger,"modInteger")
-    , (EqualsInteger,"equalsInteger")
-    , (LessThanInteger,"lessThanInteger")
-    , (LessThanEqualsInteger,"lessThanEqualsInteger")
-    , (AppendByteString,"appendByteString")
-    , (ConsByteString,"consByteString")
-    , (SliceByteString,"sliceByteString")
-    , (LengthOfByteString,"lengthOfByteString")
-    , (IndexByteString,"indexByteString")
-    , (EqualsByteString,"equalsByteString")
-    , (LessThanByteString,"lessThanByteString")
-    , (LessThanEqualsByteString,"lessThanEqualsByteString")
-    , (Sha2_256,"sha2_256")
-    , (Sha3_256,"sha3_256")
-    , (Blake2b_256,"blake2b_256")
-    , (VerifySignature,"verifySignature")
-    , (AppendString,"appendString")
-    , (EqualsString,"equalsString")
-    , (EncodeUtf8,"encodeUtf8")
-    , (DecodeUtf8,"decodeUtf8")
-    , (IfThenElse,"ifThenElse")
-    , (ChooseUnit,"chooseUnit")
-    , (Trace,"trace")
-    , (FstPair,"fstPair")
-    , (SndPair,"sndPair")
-    , (ChooseList,"chooseList")
-    , (MkCons,"mkCons")
-    , (HeadList,"headList")
-    , (TailList,"tailList")
-    , (NullList,"nullList")
-    , (ChooseData,"chooseData")
-    , (ConstrData,"constrData")
-    , (MapData,"mapData")
-    , (ListData,"listData")
-    , (IData,"iData")
-    , (BData,"bData")
-    , (UnConstrData,"unConstrData")
-    , (UnMapData,"unMapData")
-    , (UnListData,"unListData")
-    , (UnIData,"unIData")
-    , (UnBData,"unBData")
-    , (EqualsData,"equalsData")
-    , (SerialiseData,"serialiseData")
-    , (MkPairData,"mkPairData")
-    , (MkNilData,"mkNilData")
-    , (MkNilPairData,"mkNilPairData")
-    ]
+cachedBuiltin :: Map T.Text DefaultFun
+cachedBuiltin = mkCachedBuiltin [minBound .. maxBound] empty
 
+-- | Parser for builtin functions. Atm the parser can only parse `DefaultFun`.
 builtinFunction :: Parser DefaultFun
-builtinFunction =
-    choice $
-        map
-            (try . (\(fn, text) -> fn <$ symbol text))
-            builtinFnList
+builtinFunction = lexeme $ do
+    txt <- takeWhileP (Just "identifier") isIdentifierChar
+    case lookup txt cachedBuiltin of
+        Nothing      -> error $ "Not a builtin" <> show cachedBuiltin <> show txt
```

Comment:
You shouldn't call `error` here! There's nothing exceptional about a bulitin not being recognized. The `Parser` monad allows you to throw an error in it. I think the right function is [`customFailure`](https://hackage.haskell.org/package/megaparsec-9.2.0/docs/Text-Megaparsec.html#v:customFailure), but not sure.

Reply:
Oh, yeah, the error to throw is `UnknownBuiltinFunction` from `ParserError`.

Reply:
Yes, thank you! I was rushing to merge the rest so that the blocked PR can go through. I want to show the `cachedBuiltin` somehow, I think it's useful to see. I'll work on that today.